native/libcst/src/tokenizer/whitespace_parser.rs

```diff
@@ -74,12 +74,44 @@ impl<'a> Config<'a> {
                 break;
             }
         }
-        let default_newline =
-            NEWLINE_RE_2.with(|r| r.find(input).map(|m| m.as_str()).unwrap_or("\n"));
+
+        let mut lines = Vec::new();
+        let mut start = 0;
+        let mut newline_positions = memchr2_iter(b'\n', b'\r', input.as_bytes());
+
+        while let Some(newline_position) = newline_positions.next() {
+            let newline_character = input.as_bytes()[newline_position] as char;
+
+            let len = if newline_character == '\r'
+                && input.as_bytes().get(newline_position + 1) == Some(&b'\n')
+            {
+                // Skip the next '\n'
+                newline_positions.next();
+                2
+            } else {
+                1
+            };
+
+            let end = newline_position + len;
+            lines.push(&input[start..end]);
+            start = end;
+        }
+
+        // Push the last line if it isn't terminated by a newline character
+        if start < input.len() {
+            lines.push(&input[start..]);
+        }
+
+        let default_newline = match lines.first().map(|line| line.as_bytes()).unwrap_or(&[]) {
+            [.., b'\r', b'\n'] => "\r\n",
+            [.., b'\n'] => "\n",
+            [.., b'\r'] => "\r",
+            _ => "\n",
+        };
 
         Self {
             input,
-            lines: input.split_inclusive(default_newline).collect(),
+            lines,
```

Comment:
`has_trailing_newline` may need fixing too, to correctly handle `\r` endings but I'll leave this for another PR